src/newton_solver.cpp

```diff
@@ -255,22 +256,46 @@ void NewtonSolverSparse::prepareLinearSystem(int ntry, int nnewt) {
     model->fJSparse(*t, 0.0, x, &dx, &xdot, Jtmp);
 
     /* Get factorization of sparse Jacobian */
-    if(symbolic) /* if symbolic was already created free first to avoid memory leak */
-        klu_free_symbolic(&symbolic, &common);
-    symbolic = klu_analyze(model->nx_solver, Jtmp->indexptrs,
-                           Jtmp->indexvals, &common);
+    if (!symbolic) /* we only need to perform symbolic factorization once */
+        symbolic = klu_analyze(model->nx_solver, Jtmp->indexptrs,
+                               Jtmp->indexvals, &common);
+
     if (!symbolic) {
-        throw NewtonFailure(common.status,"klu_analyze");
+        throw NewtonFailure(common.status, "klu_analyze");
     }
 
-    if(numeric) /* if numeric was already created free first to avoid memory leak */
-        klu_free_numeric(&numeric, &common);
-    numeric = klu_factor(Jtmp->indexptrs, Jtmp->indexvals,
-                         Jtmp->data, symbolic, &common);
-    if (!numeric) {
-        throw NewtonFailure(common.status,"klu_factor");
+    if (numeric) { /* if numeric we only need to refactor, which can be done
+                    very effectively using klu_refactor, see cvode_klu.c for
+                    reference for this code*/
+        if (!klu_refactor(Jtmp->indexptrs, Jtmp->indexvals, Jtmp->data,
+                          symbolic, numeric, &common))
+            throw NewtonFailure(common.status, "klu_refactor");
+        /* check cheap estimate of rcond to see if we need to recompute
+         factorization
+         */
+        if (!klu_rcond(symbolic, numeric, &common))
+            throw NewtonFailure(common.status, "klu_rcond");
+
+        if (common.rcond < SUNRpowerR(UNIT_ROUNDOFF, 2 / 3)) {
```

Comment:
Too late, but: `2/3` performs integer division. The result might not be as expected. Is this intended or did you mean `2.0/3.0`? (same 5 lines below)

Reply:
duh. should be `2.0/3.0`